apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl

```diff
@@ -0,0 +1,191 @@
+%% @end
+%%--------------------------------------------------------------------
+handle_call({apply, _,_,_}, _From, #state{worker=undefined}=State) ->
+    {reply, {error, not_connected}, State};
+handle_call({apply, Mod, Fun, Args}, _From, #state{worker={_MRef, Pid}}=State) when is_pid(Pid) ->
+    Reply =
+        try
+            erlang:apply(Mod, Fun, [Pid|Args])
+        catch
+            exit:{noproc,_Reason} ->
+                %% We remove the worker pid only in handle_info where
+                %% the attempt to restart the process will also be
+                %% kicked off.
+                {error, not_connected}
+        end,
+    {reply, Reply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling cast messages
+%%
+%% @spec handle_cast(Msg, State) -> {noreply, State} |
+%%                                  {noreply, State, Timeout} |
+%%                                  {stop, Reason, State}
+%% @end
+%%--------------------------------------------------------------------
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling all non call/cast messages
+%%
+%% @spec handle_info(Info, State) -> {noreply, State} |
+%%                                   {noreply, State, Timeout} |
+%%                                   {stop, Reason, State}
+%% @end
+%%--------------------------------------------------------------------
+handle_info(connect, #state{start_mfa={M,F,A}, worker=undefined} = State) ->
+    NewState =
+    case apply(M,F,[A]) of
+        {ok, Pid} ->
+            MRef = monitor(process, Pid),
+            State#state{worker={MRef, Pid}};
+        {error, Reason} ->
+            lager:warning("Could not connect due to ~p", [Reason]),
+            schedule_reconnect(State),
+            State
+    end,
+    {noreply, NewState};
+handle_info({'DOWN', MRef, process, WorkerPid, _}, #state{worker={MRef, WorkerPid}}=S) ->
+    schedule_reconnect(S),
+    {noreply, S#state{worker=undefined}};
+handle_info({'EXIT',_,_}, State) ->
+    %% Ignore exit signals, we handle process lifecycle using
+    %% monitors. We'll receive these depending on if the start
+    %% function configured in `start_mfa` links or not.
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+%%
+%% @spec terminate(Reason, State) -> void()
+%% @end
+%%--------------------------------------------------------------------
+terminate(_Reason, #state{worker=undefined}) ->
+    ok;
+terminate(_Reason, #state{worker={Mref, Pid}}) ->
+    mc_worker:disconnect(Pid),
```

Comment:
Hmm, shouldn't this be implementation agnostic.. But for just MongoDB it's ok.

Reply:
Yeah, that's just silly. Hmmm... I guess we'd need to pass in a termination function as well so we can be sure to clean up and not leave any leaked processes behind.